### ISSUE TITLE
add 'strict' mode to user's permset config

### ISSUE THREAD

**Issue by dawiddiwad** [dawiddiwad/STAF#31]
Right now permset config is inclusive. Meaning it can contain extra permissions apart from those specified in the user's config.

Add an option to make it strict (exclusive), for example, a flag on the user's config:
![image](https://github.com/dawiddiwad/STAF/assets/40611553/1fdf8c04-c5ec-4832-ba91-04a621c0cb91)

**Comment by dawiddiwad** [comment]
```sql
SELECT Id, Username
FROM User
WHERE IsActive = true
    AND UserType = 'Standard'
    AND Profile.Name = 'Oxagon Standard User'
    AND UserRole.Name = 'Port Executive Director'
    AND LocaleSidKey = 'en_AE'
    AND Id IN (
        SELECT AssigneeId
        FROM PermissionSetAssignment
        WHERE IsActive = true
            AND PermissionSet.Name = 'PortSalesDirectorGroup'
    )
    AND Id NOT IN (
        SELECT AssigneeId
        FROM PermissionSetAssignment
        WHERE IsActive = true
            AND PermissionSet.Name != 'PortSalesDirectorGroup'
            AND PermissionSet.ProfileId = null
    )
```

**Comment by dawiddiwad** [comment]
also fix the 
```EXCEEDED_MAX_SEMIJOIN_SUBSELECTS: Maximum 2 semi join sub-selects are allowed```
limit

**Comment by dawiddiwad** [comment]
another idea use `GROUP BY` `SELECT COUNT_DISTINCT` `HAVING` keywords combined with query like this:
![image](https://github.com/dawiddiwad/STAF/assets/40611553/ac128c05-771f-4f6a-9e83-4ccdd85d3570)

**Comment by dawiddiwad** [comment]
permset list version:
```sql
SELECT Assignee.Username, AssigneeId
FROM PermissionSetAssignment
WHERE IsActive = true
    AND Assignee.UserType = 'Standard'
    AND Assignee.Profile.Name = 'Oxagon Standard User'
    AND Assignee.LocaleSidKey = 'en_AE'
AND PermissionSet.Name IN ('PortSalesManagerGroup', 'HarbormasterGroup', 'PSWInternalUser')
GROUP BY Assignee.Username, AssigneeId 
HAVING COUNT(Assignee.Username) = 3
```

without permset list version:
```sql
SELECT Assignee.Username, AssigneeId
FROM PermissionSetAssignment
WHERE IsActive = true
    AND Assignee.UserType = 'Standard'
    AND Assignee.Profile.Name = 'Oxagon Standard User'
    AND Assignee.LocaleSidKey = 'en_AE'
GROUP BY Assignee.Username, AssigneeId 
HAVING COUNT(Assignee.Username) = 1
```

**Comment by dawiddiwad** [comment]
Still occurs, fix is to query for all PermissionSetAssignment based on user config fields and then filter cached results.

example SOQL:
```sql
SELECT AssigneeId, Assignee.Name, PermissionSet.Name
FROM PermissionSetAssignment
WHERE IsActive = true
    AND Assignee.UserType = 'Standard'
    AND Assignee.Profile.Name = 'Oxagon Standard User'
    AND Assignee.LocaleSidKey = 'en_AE'
Order by Assignee.Name desc
```

example code to get matching users:
```typescript
const permissionSetConfig = ['ICCUser', 'SEDO_SharedPermissionsGroup']

const includesAll = (source: string[], matching: string[]) => matching.every(value => source.includes(value));
const includesSome = (source: string[], matching: string[]) => matching.some(value => source.includes(value));

const matchingUsers = new Set<string>();
queryResults.forEach(candidate => {
    const candidateAssignments = queryResults
        .filter(target => target.AssigneeId === candidate.AssigneeId)
        .map(target => target.PermissionSet.Name);
    
    if (
        (config.permissionSets
        && config.strictPermissionSets
        && includesAll(candidateAssignments, permissionSetConfig)
        && candidateAssignments.length === permissionSetConfig.length + 1)
        || (config.permissionSets
        && includesSome(permissionSetAssignments, permissionSetConfig))
        || (!config.permissionSets 
        && config.strictPermissionSets
        && candidateAssignments.length === 1)
        || (!config.permissionSets 
        && !config.strictPermissionSets)
    ) {
        matchingUsers.add(candidate.AssigneeId);
    }
});

console.log([...matchingUsers]);
```

**Comment by dawiddiwad** [comment]
closed in #35
